be/src/fs/hdfs/fs_hdfs.cpp

```diff
@@ -50,58 +179,48 @@ class HdfsInputStream : public io::SeekableInputStream {
     void set_size(int64_t size) override;
 
 private:
-    hdfsFS _fs;
-    hdfsFile _file;
-    std::string _file_name;
+    std::unique_ptr<GetHdfsFileReadOnlyHandle> _handle;
     int64_t _offset{0};
     int64_t _file_size{0};
 };
 
 HdfsInputStream::~HdfsInputStream() {
     auto ret = call_hdfs_scan_function_in_pthread([this]() {
-        int r = hdfsCloseFile(this->_fs, this->_file);
+        int r = _handle->close();
         if (r == -1) {
-            auto error_msg = fmt::format("Fail to close file {}: {}", _file_name, get_hdfs_err_msg());
+            auto error_msg = fmt::format("Fail to close file {}: {}", _handle->getPath(), get_hdfs_err_msg());
             LOG(WARNING) << error_msg;
             return Status::IOError(error_msg);
         }
         return Status::OK();
     });
     Status st = ret->get_future().get();
-    PLOG_IF(ERROR, !st.ok()) << "close " << _file_name << " failed";
+    PLOG_IF(ERROR, !st.ok()) << "close " << _handle->getPath() << " failed";
 }
 
 StatusOr<int64_t> HdfsInputStream::read(void* data, int64_t size) {
     if (UNLIKELY(size > std::numeric_limits<tSize>::max())) {
         size = std::numeric_limits<tSize>::max();
     }
-    tSize r = hdfsPread(_fs, _file, _offset, data, static_cast<tSize>(size));
-    if (r == -1) {
-        return Status::IOError(fmt::format("fail to hdfsPread {}: {}", _file_name, get_hdfs_err_msg()));
-    }
-    _offset += r;
-    return r;
+    return _handle->pread(static_cast<uint8_t*>(data), size, config::hdfs_client_io_read_retry);
+    // return _handle->read(static_cast<uint8_t*>(data), size, config::hdfs_client_io_read_retry);
 }
 
 Status HdfsInputStream::seek(int64_t offset) {
     if (offset < 0) return Status::InvalidArgument(fmt::format("Invalid offset {}", offset));
-    _offset = offset;
+    _handle->setOffset(offset);
     return Status::OK();
 }
 
 StatusOr<int64_t> HdfsInputStream::get_size() {
     if (_file_size == 0) {
         auto ret = call_hdfs_scan_function_in_pthread([this] {
-            auto info = hdfsGetPathInfo(_fs, _file_name.c_str());
-            if (UNLIKELY(info == nullptr)) {
-                return Status::IOError(fmt::format("Fail to get path info of {}: {}", _file_name, get_hdfs_err_msg()));
-            }
-            this->_file_size = info->mSize;
-            hdfsFreeFileInfo(info, 1);
+            auto st = _handle->getSize();
```

Comment:
if `_handle` only return `_fs`, maybe we don't change lot of codes, what do you think